File: n653-tow-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findTarget(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: bool
        """
        data = {}
        return self.dfs(root, k, data)
    
    def dfs(self, root, k, data):
        if not root:
            return False
        
        if k - root.val in data:
            return True
        
        data[root.val] = 1
        
        return self.dfs(root.left, k, data) or self.dfs(root.right, k, data)
        

File: n653-tow-sum-iv-input-is-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
var findTarget = function(root, k) {
    let data = {}
    
    // 闭包，可以直接访问外层函数的变量
    function dfs(root) {
        if (!root) return false;
        
        if (data[k-root.val]) return true;
        
        data[root.val] = 1
        return dfs(root.left) || dfs(root.right)
    }
     
    return dfs(root)
};
